File: src/searching/binarySearch.js
/**
 * BINARY SEARCH
 * - Takes in sorted/unsorted) array
 * - Takes in item to be searched
 * - Sorts the array
 * - Applies binary search
 * */

const ARRAY = [4, 7, 8, 45436, 345, 367, 22, 1, 99, 34, 2, 9, 11, 17, 99];
const TO_FIND = 99;

const binarySearch = (list, item) => {
  let lowerBound = 0;
  let upperBound = list.length - 1;

  while (lowerBound <= upperBound) {
    let middleBound = Math.floor((lowerBound + upperBound) / 2);
    if (list[middleBound] < item) lowerBound = middleBound + 1;  // ±1 because we've already checked the middleBound
    else if (list[middleBound] > item) upperBound = middleBound - 1;
    else if (list[middleBound] === item) return `${item} found at index ${middleBound}`;
    else return `${item} not found`;
  }

};

const sortedArray = ARRAY.sort((a, b) => a > b ? 1 : (a < b ? -1 : 0));  // added () for readability
console.log(binarySearch(sortedArray, TO_FIND));

File: src/searching/binary_search.py
def binary_search(arr, num_to_find):
    lower_bound = 0
    upper_bound = len(arr) - 1
    
    while lower_bound <= upper_bound:
        index_avg = (lower_bound + upper_bound) // 2
        middle_element = arr[index_avg]
        if middle_element == num_to_find:
            return index_avg
        elif middle_element < num_to_find:
            lower_bound = index_avg + 1
        else:
            upper_bound = index_avg - 1
    return None


print(binary_search([1, 5, 8, 11, 14], 7))
